src/automatic.ts

```diff
@@ -192,19 +206,31 @@ export class AutomaticMode {
     const {driver, initialTabHandle} =
         browsers.get(browserSignature(spec.browser))!;
 
-    let millis: number|undefined;
     let bytesSent = 0;
     let userAgent = '';
     // TODO(aomarks) Make maxAttempts and timeouts configurable.
     const maxAttempts = 3;
+    const measurements = spec.measurement;
+    let millis: number[];
+    let numPending: number;
     for (let attempt = 1;; attempt++) {
+      millis = [];
+      numPending = measurements.length;
       await openAndSwitchToNewTab(driver, spec.browser);
       await driver.get(url);
-      for (let waited = 0; millis === undefined && waited <= 10000;
-           waited += 50) {
+      for (let waited = 0; numPending > 0 && waited <= 10000; waited += 50) {
         // TODO(aomarks) You don't have to wait in callback mode!
         await wait(50);
-        millis = await measure(driver, spec.measurement, server);
+        for (let i = 0; i < measurements.length; i++) {
```

Comment:
With this level of nested loops and conditionals, I'm finding it a bit hard to follow. Some more comments would be nice and probably some more descriptive variable names. Renaming `i` to something like `measurementIndex` would let the reader more easily see `millis` is indexed by measurement. A new name for `millis` might help (`measurementResults`?), `numPending` could be `pendingMeasurementCount`, etc. You could see if you can address this in one of the next PRs too.

Reply:
Done, renamed a bunch of stuff, removed a few variables. Still could probably be refactored to be more clear, but hopefully this is a good enough improvement for now.